### PR TITLE
[Java] Fix Kryo FP & Kryo 5 Support

### LINKED ISSUE #4992: LGTM.com - false positive: Java: Kryo from KryoPool.Builder
**Description of the false positive**

The Kryo library doesn't correctly track a safe instance of `Kryo` being passed to `KryoPool.Builder`.

As an example, this should be considered safe:
```java
private final KryoPool kryoPool = new KryoPool.Builder(this::newKryo).softReferences().build();

private Kryo newKryo() {
    Kryo kryo = new Kryo();
    kryo.setRegistrationRequired(true);
    // ... ect...
	return kryo;
}

private T readFromInput(Input input) {
	return kryoPool.run(kryo -> kryo.readClassAndObject(input));
}
```

### PR DESCRIPTION
**Pull request by JLLeitschuh**
Closes #4992

### CODE REVIEW

File: java/change-notes/2021-05-05-kryo-improvements.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* Add support for version 5 of the Kryo serialzation/deserialization framework.
+* Add support for detecting safe uses of Kryo utilizing `KryoPool.Builder. (#4992)[https://github.com/github/codeql/issues/4992]
```
Comment (Marcono1234) [inline java/change-notes/2021-05-05-kryo-improvements.md:3]:
```suggestion
* Add support for detecting safe uses of Kryo utilizing `KryoPool.Builder`. [#4992](https://github.com/github/codeql/issues/4992)
```

File: java/ql/src/semmle/code/java/frameworks/Kryo.qll [hunk 2]
```diff
@@ -3,19 +3,60 @@
  */
 
 import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.FlowSteps
 
 /**
  * The type `com.esotericsoftware.kryo.Kryo`.
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/frameworks/Kryo.qll:10]: Should the other package name be mentioned here as well, or should the package name could just be removed?

File: java/ql/src/semmle/code/java/frameworks/Kryo.qll [hunk 3]
```diff
@@ -3,19 +3,60 @@
...
+ */
+class KryoPoolBuilder extends RefType {
+  KryoPoolBuilder() {
+    hasQualifiedName("com.esotericsoftware.kryo.pool", "KryoPool$Builder") or
+    hasQualifiedName("com.esotericsoftware.kryo5.pool", "KryoPool$Builder")
+  }
+}
+
+/**
+ * A Kryo pool builder method used a fluent API call chain.
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/frameworks/Kryo.qll:50]:
```suggestion
 * A Kryo pool builder method used in a fluent API call chain.
```

File: java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll [hunk 4]
```diff
@@ -48,6 +48,65 @@ class SafeKryo extends DataFlow2::Configuration {
...
+
+  override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    stepKryoPoolBuilderFactoryArgToConstructor(node1, node2) or
+    stepKryoPoolRunMethodAccessQualifierToFunctionalArgument(node1, node2) or
+    stepKryoPoolBuilderChainMethod(node1, node2) or
+    stepKryoPoolBorrowMethod(node1, node2)
+  }
+
+  /**
+   * Holds when a funcitonal expression is used to create a `KryoPool.Builder`.
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/security/UnsafeDeserialization.qll:60]:
```suggestion
   * Holds when a functional expression is used to create a `KryoPool.Builder`.
```

File: java/ql/src/semmle/code/java/frameworks/Kryo.qll (hunk 2 above)
Comment (Marcono1234) [reply java/ql/src/semmle/code/java/frameworks/Kryo.qll:10]: For reference, the `kryo` and `kryo5` packages seem to contain exactly the same content, see https://github.com/EsotericSoftware/kryo/pull/652.

File: java/change-notes/2021-05-05-kryo-improvements.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+lgtm,codescanning
+* Add support for version 5 of the Kryo serialzation/deserialization framework.
```
[the remark this answers is not in the document]
Comment (JLLeitschuh) [reply java/change-notes/2021-05-05-kryo-improvements.md]: Thanks!

**Review by aschackmull — APPROVED** [review]

**Pull request merged**